src/prop/proof_cnf_stream.cpp

```diff
@@ -0,0 +1,1265 @@
+                        "MACRO_SR_PRED_TRANSFORM added norm "
+                     << node[0] << "\n";
+      }
+      convertAndAssert(node[0], !negated);
+      break;
+    }
+    case kind::EQUAL:
+      if (node[0].getType().isBoolean())
+      {
+        convertAndAssertIff(node, negated);
+        break;
+      }
+      CVC4_FALLTHROUGH;
+    default:
+    {
+      // negate
+      Node nnode = negated ? node.negate() : static_cast<Node>(node);
+      // Atoms
+      SatLiteral lit = toCNF(node, negated);
+      bool added = d_cnfStream.assertClause(nnode, lit);
+      if (negated && added && nnode != node.notNode())
+      {
+        d_proof.addStep(
+            nnode, PfRule::MACRO_SR_PRED_TRANSFORM, {node.notNode()}, {nnode});
+        Trace("cnf") << "ProofCnfStream::convertAndAssert: "
+                        "MACRO_SR_PRED_TRANSFORM added norm "
+                     << nnode << "\n";
+      }
+      if (added)
+      {
+        MinisatSatSolver* minisat =
+            static_cast<MinisatSatSolver*>(d_cnfStream.d_satSolver);
+        minisat->getProofManager()->registerSatAssumptions({nnode});
+      }
+    }
+  }
+}
+
+void ProofCnfStream::convertAndAssertAnd(TNode node, bool negated)
+{
+  Trace("cnf") << "ProofCnfStream::convertAndAssertAnd(" << node
+               << ", negated = " << (negated ? "true" : "false") << ")\n";
+  Assert(node.getKind() == kind::AND);
+  if (!negated)
+  {
+    // If the node is a conjunction, we handle each conjunct separately
+    NodeManager* nm = NodeManager::currentNM();
+    for (unsigned i = 0, size = node.getNumChildren(); i < size; ++i)
+    {
+      // Create a proof step for each n_i
+      Node iNode = nm->mkConst<Rational>(i);
+      d_proof.addStep(node[i], PfRule::AND_ELIM, {node}, {iNode});
+      Trace("cnf") << "ProofCnfStream::convertAndAssertAnd: AND_ELIM " << i
+                   << " added norm " << node[i] << "\n";
+      convertAndAssert(node[i], false);
+    }
+  }
+  else
+  {
+    // If the node is a disjunction, we construct a clause and assert it
+    unsigned i, size = node.getNumChildren();
+    SatClause clause(size);
+    for (i = 0; i < size; ++i)
+    {
+      clause[i] = toCNF(node[i], true);
+    }
+    bool added = d_cnfStream.assertClause(node.negate(), clause);
+    // register proof step
+    if (added)
+    {
+      std::vector<Node> disjuncts;
+      for (i = 0; i < size; ++i)
+      {
+        disjuncts.push_back(node[i].notNode());
+      }
+      Node clauseNode = NodeManager::currentNM()->mkNode(kind::OR, disjuncts);
+      d_proof.addStep(clauseNode, PfRule::NOT_AND, {node.notNode()}, {});
+      Trace("cnf") << "ProofCnfStream::convertAndAssertAnd: NOT_AND added "
+                   << clauseNode << "\n";
+      // justify normalized clause as well, since that's what will be saved in
+      // the SAT solver and registered with prop engine
+      Node normClauseNode = d_psb.factorReorderElimDoubleNeg(clauseNode);
```

Comment:
It seems that there is an invariant that this method should be called whenever you registerSatAssumptions below.

Would it make sense to make a utility function that does this?  It seems this pattern is repeated several times below.

Reply:
Indeed. I've updated the class to do this.